mteb/abstasks/AbsTaskClusteringFast.py

```diff
@@ -0,0 +1,189 @@
+        n_init="auto",
+    )
+
+    for _ in range(n_clusters):
+        # sample N samples from the corpus without replacement
+        cluster_indices = rng_state.sample(range(n_embeddings), k=cluster_size)
+
+        _embeddings = embeddings[cluster_indices]
+        _labels = labels_arr[cluster_indices]
+        clustering_model.fit(_embeddings)
+        cluster_assignment = clustering_model.labels_
+        v_measure = v_measure_score(_labels, cluster_assignment)
+        scores.append({"v_measure": v_measure})
+
+    return scores
+
+
+class AbsTaskClusteringFast(AbsTask):
+    """Abstract class for Clustering tasks.
+
+    This class embeds the corpus sentences then samples N samples from the corpus and clusters them.
+    he similarity then is calculated using the V-measure metric, which is invariant to the permutation of the labels.
+    This approach is then repeated K times.
+
+    self.load_data() must generate a huggingface dataset with a split matching self.metadata_dict["eval_splits"], and assign it to self.dataset.
+    It must contain the following columns:
+        sentences: list of str
+        labels: list of str
+    """
+
+    max_documents_to_embed = 16_384
+    max_documents_per_cluster = 2048
+    n_clusters = 10
+    k_mean_batch_size = 512
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def calculate_main_score(self, scores: list[dict[str, float]]) -> float:
+        return float(np.mean([score[self.metadata.main_score] for score in scores]))
+
+    def evaluate(
+        self, model, split="test", **kwargs
+    ) -> (
+        dict[HFLang, dict[str, ClusteringScores | float]]
+        | dict[str, ClusteringScores | float]
+    ):
+        self.dataset: MultilingualDataset | HFDataset
+
+        if not self.data_loaded:
+            self.load_data()
+
+        if self.is_multilingual:
+            multilingual_ds: MultilingualDataset = self.dataset  # type: ignore
+
+            multilingual_scores: dict[HFLang, dict[str, ClusteringScores | float]] = {}
+            for lang in self.dataset:
+                logger.info(
+                    f"\nTask: {self.metadata.name}, split: {split}, language: {lang}. Running..."
+                )
+                multilingual_scores[lang] = self._evaluate_monolingual(
+                    model, multilingual_ds[lang], split, **kwargs
+                )
+                return multilingual_scores
+        logger.info(f"\nTask: {self.metadata.name}, split: {split}. Running...")
+
+        ds: HFDataset = self.dataset  # type: ignore
+        scores = self._evaluate_monolingual(model, ds, split, **kwargs)
+        return scores
+
+    def _evaluate_monolingual(
+        self, model, dataset: HFDataset, split: str = "test", **kwargs: Any
+    ) -> dict[str, float | ClusteringScores]:
+        sentences = dataset[split]["sentences"][: self.max_documents_to_embed]
+        labels = dataset[split]["labels"][: self.max_documents_to_embed]
```

Comment:
Right now it always takes the first `self.max_documents_to_embed` samples to encode. Should we shuffle those first?

Reply:
Good question! 

I see two potential solutions:

1) sample randomly as you suggest
2) take the first n., this will allow for streaming datasets

Either solution if fine with me (It should def. not be the way it is now)

Reply:
I went with 1)